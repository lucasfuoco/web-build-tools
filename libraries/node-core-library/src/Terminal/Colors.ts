// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

/**
 * @beta
 */
export interface IColorableSequence {
  text: string;
  isEol?: boolean;
  foregroundColor?: ColorValue;
  backgroundColor?: ColorValue;
}

export const eolSequence: IColorableSequence = {
  isEol: true
} as IColorableSequence;

/**
 * @beta
 */
export enum ColorValue {
  Black,
  Red,
  Green,
  Yellow,
  Blue,
  Magenta,
  Cyan,
  White,
  Gray
}

/**
 * The static functions on this class are used to produce colored text
 * for use with the node-core-library terminal.
 *
 * @example
 * terminal.writeLine(Colors.green('Green Text!'), ' ', Colors.blue('Blue Text!'));
 *
 * @beta
 */
export class Colors {
  public static black(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Black
    };
  }

  public static red(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Red
    };
  }

  public static green(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Green
    };
  }

  public static yellow(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Yellow
    };
  }

  public static blue(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Blue
    };
  }

  public static magenta(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Magenta
    };
  }

  public static cyan(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Cyan
    };
  }

  public static white(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.White
    };
  }

  public static gray(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      foregroundColor: ColorValue.Gray
    };
  }

  public static blackBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Black
    };
  }

  public static redBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Red
    };
  }

  public static greenBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Green
    };
  }

  public static yellowBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Yellow
    };
  }

  public static blueBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Blue
    };
  }

  public static magentaBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Magenta
    };
  }

  public static cyanBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Cyan
    };
  }

  public static whiteBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.White
    };
  }

  public static grayBackground(text: string | IColorableSequence): IColorableSequence {
    return {
      ...Colors._normalizeStringOrColorableSequence(text),
      backgroundColor: ColorValue.Gray
    };
  }

  /**
   * @internal
   */
  public static _normalizeStringOrColorableSequence(value: string | IColorableSequence): IColorableSequence {
    if (typeof value === 'string') {
      return {
        text: value
      };
    } else {
      return value;
    }
  }
}