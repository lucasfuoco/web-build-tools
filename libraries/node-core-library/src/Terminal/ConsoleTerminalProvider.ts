// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { EOL } from 'os';
import { enabled as supportsColor } from 'colors/safe';

import { ITerminalProvider, TerminalProviderSeverity } from './ITerminalProvider';

/**
 * Options to be provided to a {@link ConsoleTerminalProvider}
 *
 * @beta
 */
export interface IConsoleTerminalProviderOptions {
  /**
   * If true, print verbose logging messages
   */
  verboseEnabled: boolean;
}

/**
 * @beta
 */
export class ConsoleTerminalProvider implements ITerminalProvider {
  public verboseEnabled: boolean = false;

  public constructor(options: Partial<IConsoleTerminalProviderOptions> = {}) {
    this.verboseEnabled = !!options.verboseEnabled;
  }

  public write(data: string, severity: TerminalProviderSeverity): void {
    switch (severity) {
      case TerminalProviderSeverity.warning:
      case TerminalProviderSeverity.error: {
        process.stderr.write(data);
        break;
      }

      case TerminalProviderSeverity.verbose: {
        if (this.verboseEnabled) {
          process.stdout.write(data);
        }
        break;
      }

      case TerminalProviderSeverity.log:
      default: {
        process.stdout.write(data);
        break;
      }
    }
  }

  public get eolCharacter(): string {
    return EOL;
  }

  public get supportsColor(): boolean {
    return supportsColor;
  }
}
