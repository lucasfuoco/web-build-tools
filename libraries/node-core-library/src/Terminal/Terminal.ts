// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import {
  ITerminalProvider,
  TerminalProviderSeverity
} from './ITerminalProvider';
import {
  IColorableSequence,
  ColorValue,
  Colors,
  eolSequence
} from './Colors';

/**
 * This class facilitates writing to a console.
 *
 * @beta
 */
export class Terminal {
  private _providers: Set<ITerminalProvider>;

  public constructor(provider: ITerminalProvider) {
    this._providers = new Set<ITerminalProvider>();
    this._providers.add(provider);
  }

  /**
   * Subscribe a new terminal provider.
   */
  public registerProvider(provider: ITerminalProvider): void {
    this._providers.add(provider);
  }

  /**
   * Unsubscribe a terminal provider. If the provider isn't subscribed, this function does nothing.
   */
  public unregisterProvider(provider: ITerminalProvider): void {
    if (this._providers.has(provider)) {
      this._providers.delete(provider);
    }
  }

  /**
   * Write a generic message to the terminal
   */
  public write(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.log);
  }

  /**
   * Write a generic message to the terminal, followed by a newline
   */
  public writeLine(...messageParts: (string | IColorableSequence)[]): void {
    this.write(...messageParts, eolSequence);
  }

  /**
   * Write a warning message to the console with yellow text.
   *
   * @remarks
   * The yellow color takes precedence over any other foreground colors set.
   */
  public writeWarning(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Yellow
        })
      ),
      TerminalProviderSeverity.warning
    );
  }

  /**
   * Write a warning message to the console with yellow text, followed by a newline.
   *
   * @remarks
   * The yellow color takes precedence over any other foreground colors set.
   */
  public writeWarningLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Yellow
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.warning
    );
  }

  /**
   * Write an error message to the console with red text.
   *
   * @remarks
   * The red color takes precedence over any other foreground colors set.
   */
  public writeError(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      messageParts.map(
        (part): IColorableSequence => ({
          ...Colors._normalizeStringOrColorableSequence(part),
          foregroundColor: ColorValue.Red
        })
      ),
      TerminalProviderSeverity.error
    );
  }

  /**
   * Write an error message to the console with red text, followed by a newline.
   *
   * @remarks
   * The red color takes precedence over any other foreground colors set.
   */
  public writeErrorLine(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(
      [
        ...messageParts.map(
          (part): IColorableSequence => ({
            ...Colors._normalizeStringOrColorableSequence(part),
            foregroundColor: ColorValue.Red
          })
        ),
        eolSequence
      ],
      TerminalProviderSeverity.error
    );
  }

  /**
   * Write a verbose-level message.
   */
  public writeVerbose(...messageParts: (string | IColorableSequence)[]): void {
    this._writeSegmentsToProviders(messageParts, TerminalProviderSeverity.verbose);
  }

  /**
   * Write a verbose-level message followed by a newline.
   */
  public writeVerboseLine(...messageParts: (string | IColorableSequence)[]): void {
    this.writeVerbose(...messageParts, eolSequence);
  }

  private _writeSegmentsToProviders(
    segments: (string | IColorableSequence)[],
    severity: TerminalProviderSeverity
  ): void {

    const withColorText: { [eolChar: string]: string } = {};
    const withoutColorText: { [eolChar: string]: string } = {};
    let withColorLines: string[] | undefined;
    let withoutColorLines: string[] | undefined;

    this._providers.forEach((provider) => {
      const eol: string = provider.eolCharacter;
      let textToWrite: string;
      if (provider.supportsColor) {
        if (!withColorLines) {
          withColorLines = this._serializeFormattableTextSegments(segments, true);
        }

        if (!withColorText[eol]) {
          withColorText[eol] = withColorLines.join(eol);
        }

        textToWrite = withColorText[eol];
      } else {
        if (!withoutColorLines) {
          withoutColorLines = this._serializeFormattableTextSegments(segments, false);
        }

        if (!withoutColorText[eol]) {
          withoutColorText[eol] = withoutColorLines.join(eol);
        }

        textToWrite = withoutColorText[eol];
      }

      provider.write(textToWrite, severity);
    });
  }

  private _serializeFormattableTextSegments(segments: (string | IColorableSequence)[], withColor: boolean): string[] {
    const lines: string[] = [];
    let segmentsToJoin: string[] = [];
    let lastSegmentWasEol: boolean = false;
    for (let i: number = 0; i < segments.length; i++) {
      const segment: IColorableSequence = Colors._normalizeStringOrColorableSequence(segments[i]);
      lastSegmentWasEol = !!segment.isEol;
      if (lastSegmentWasEol) {
        lines.push(segmentsToJoin.join(''));
        segmentsToJoin = [];
      } else {

        if (withColor) {
          const startColorCodes: number[] = [];
          const endColorCodes: number[] = [];
          switch (segment.foregroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(30);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(31);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(32);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(33);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(34);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(35);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(36);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(37);
              endColorCodes.push(39);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(90);
              endColorCodes.push(39);
              break;
            }
          }

          switch (segment.backgroundColor) {
            case ColorValue.Black: {
              startColorCodes.push(40);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Red: {
              startColorCodes.push(41);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Green: {
              startColorCodes.push(42);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Yellow: {
              startColorCodes.push(43);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Blue: {
              startColorCodes.push(44);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Magenta: {
              startColorCodes.push(45);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Cyan: {
              startColorCodes.push(46);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.White: {
              startColorCodes.push(47);
              endColorCodes.push(49);
              break;
            }

            case ColorValue.Gray: {
              startColorCodes.push(100);
              endColorCodes.push(49);
              break;
            }
          }

          for (let j: number = 0; j < startColorCodes.length; j++) {
            const code: number = startColorCodes[j];
            segmentsToJoin.push(...[
              '\u001b[',
              code.toString(),
              'm'
            ]);
          }

          segmentsToJoin.push(segment.text);

          for (let j: number = endColorCodes.length - 1; j >= 0; j--) {
            const code: number = endColorCodes[j];
            segmentsToJoin.push(...[
              '\u001b[',
              code.toString(),
              'm'
            ]);
          }
        } else {
          segmentsToJoin.push(segment.text);
        }
      }
    }

    if (segmentsToJoin.length > 0) {
      lines.push(segmentsToJoin.join(''));
    }

    if (lastSegmentWasEol) {
      lines.push('');
    }

    return lines;
  }
}
