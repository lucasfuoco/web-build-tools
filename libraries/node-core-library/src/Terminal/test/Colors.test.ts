// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { Terminal } from '../Terminal';
import { StringBufferTerminalProvider } from './StringBufferTerminalProvider';
import { createColorGrid } from './createColorGrid';

describe('Colors', () => {
  let terminal: Terminal;
  let provider: StringBufferTerminalProvider;

  beforeEach(() => {
    provider = new StringBufferTerminalProvider(true);
    terminal = new Terminal(provider);
  });

  test('writes color grid correctly', () => {
    for (const line of createColorGrid()) {
      terminal.writeLine(...line);
    }

    expect(provider.getOutput()).toMatchSnapshot();
  });
});
