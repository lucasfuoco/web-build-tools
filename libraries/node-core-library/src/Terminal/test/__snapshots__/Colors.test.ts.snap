// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colors writes color grid correctly 1`] = `"X[x][30mX[x][39m[x][37mX[x][39m[x][90mX[x][39m[x][35mX[x][39m[x][31mX[x][39m[x][33mX[x][39m[x][32mX[x][39m[x][36mX[x][39m[x][34mX[x][39m[n][x][40mX[x][49m[x][30m[x][40mX[x][49m[x][39m[x][37m[x][40mX[x][49m[x][39m[x][90m[x][40mX[x][49m[x][39m[x][35m[x][40mX[x][49m[x][39m[x][31m[x][40mX[x][49m[x][39m[x][33m[x][40mX[x][49m[x][39m[x][32m[x][40mX[x][49m[x][39m[x][36m[x][40mX[x][49m[x][39m[x][34m[x][40mX[x][49m[x][39m[n][x][47mX[x][49m[x][30m[x][47mX[x][49m[x][39m[x][37m[x][47mX[x][49m[x][39m[x][90m[x][47mX[x][49m[x][39m[x][35m[x][47mX[x][49m[x][39m[x][31m[x][47mX[x][49m[x][39m[x][33m[x][47mX[x][49m[x][39m[x][32m[x][47mX[x][49m[x][39m[x][36m[x][47mX[x][49m[x][39m[x][34m[x][47mX[x][49m[x][39m[n][x][100mX[x][49m[x][30m[x][100mX[x][49m[x][39m[x][37m[x][100mX[x][49m[x][39m[x][90m[x][100mX[x][49m[x][39m[x][35m[x][100mX[x][49m[x][39m[x][31m[x][100mX[x][49m[x][39m[x][33m[x][100mX[x][49m[x][39m[x][32m[x][100mX[x][49m[x][39m[x][36m[x][100mX[x][49m[x][39m[x][34m[x][100mX[x][49m[x][39m[n][x][45mX[x][49m[x][30m[x][45mX[x][49m[x][39m[x][37m[x][45mX[x][49m[x][39m[x][90m[x][45mX[x][49m[x][39m[x][35m[x][45mX[x][49m[x][39m[x][31m[x][45mX[x][49m[x][39m[x][33m[x][45mX[x][49m[x][39m[x][32m[x][45mX[x][49m[x][39m[x][36m[x][45mX[x][49m[x][39m[x][34m[x][45mX[x][49m[x][39m[n][x][41mX[x][49m[x][30m[x][41mX[x][49m[x][39m[x][37m[x][41mX[x][49m[x][39m[x][90m[x][41mX[x][49m[x][39m[x][35m[x][41mX[x][49m[x][39m[x][31m[x][41mX[x][49m[x][39m[x][33m[x][41mX[x][49m[x][39m[x][32m[x][41mX[x][49m[x][39m[x][36m[x][41mX[x][49m[x][39m[x][34m[x][41mX[x][49m[x][39m[n][x][43mX[x][49m[x][30m[x][43mX[x][49m[x][39m[x][37m[x][43mX[x][49m[x][39m[x][90m[x][43mX[x][49m[x][39m[x][35m[x][43mX[x][49m[x][39m[x][31m[x][43mX[x][49m[x][39m[x][33m[x][43mX[x][49m[x][39m[x][32m[x][43mX[x][49m[x][39m[x][36m[x][43mX[x][49m[x][39m[x][34m[x][43mX[x][49m[x][39m[n][x][42mX[x][49m[x][30m[x][42mX[x][49m[x][39m[x][37m[x][42mX[x][49m[x][39m[x][90m[x][42mX[x][49m[x][39m[x][35m[x][42mX[x][49m[x][39m[x][31m[x][42mX[x][49m[x][39m[x][33m[x][42mX[x][49m[x][39m[x][32m[x][42mX[x][49m[x][39m[x][36m[x][42mX[x][49m[x][39m[x][34m[x][42mX[x][49m[x][39m[n][x][46mX[x][49m[x][30m[x][46mX[x][49m[x][39m[x][37m[x][46mX[x][49m[x][39m[x][90m[x][46mX[x][49m[x][39m[x][35m[x][46mX[x][49m[x][39m[x][31m[x][46mX[x][49m[x][39m[x][33m[x][46mX[x][49m[x][39m[x][32m[x][46mX[x][49m[x][39m[x][36m[x][46mX[x][49m[x][39m[x][34m[x][46mX[x][49m[x][39m[n][x][44mX[x][49m[x][30m[x][44mX[x][49m[x][39m[x][37m[x][44mX[x][49m[x][39m[x][90m[x][44mX[x][49m[x][39m[x][35m[x][44mX[x][49m[x][39m[x][31m[x][44mX[x][49m[x][39m[x][33m[x][44mX[x][49m[x][39m[x][32m[x][44mX[x][49m[x][39m[x][36m[x][44mX[x][49m[x][39m[x][34m[x][44mX[x][49m[x][39m[n]"`;
